### PR TITLE
Adding Puppet Actor

### LINKED ISSUE #566: Implement puppet actor for testing
**Task summary**

Not crucial, only used for testing (and honestly I don't see what benefit this gives or what specific reason for this and the tests) but just opening issue to track

**Specification reference**

- 

**Other information and links**

### PR DESCRIPTION
**Pull request by nannick**
**Summary of changes**
Changes introduced in this pull request:
-  Introduces New Puppet Actor, supposedly only used for testing
- Adds Test for Puppet actor

**Reference issue to close (if applicable)**

Closes #566 

**Other information and links**

### CODE REVIEW

File: vm/actor/src/builtin/puppet/mod.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendReturn {
+    pub return_bytes: Serialized,
+    pub code: ExitCode,
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct FailToMarshalCBOR {}
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:42]: This doesn't fail?

File: vm/actor/src/builtin/puppet/mod.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        // TODO revisit based on issue: https://github.com/filecoin-project/specs-actors/issues/317
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:63]: what does this have to do with it?

File: vm/actor/src/builtin/puppet/mod.rs (hunk 1 above)
Comment (nannick) [reply vm/actor/src/builtin/puppet/mod.rs:42]: Nope, serde really is magic

File: vm/actor/src/builtin/puppet/mod.rs (hunk 2 above)
Comment (nannick) [reply vm/actor/src/builtin/puppet/mod.rs:63]: my bad copied over the comment by accident

File: vm/actor/src/builtin/puppet/mod.rs (hunk 1 above)
Comment (austinabell) [reply vm/actor/src/builtin/puppet/mod.rs:42]: No, I mean the intention of this struct is that it _should_ fail

Comment (nannick) [reply vm/actor/src/builtin/puppet/mod.rs:42]: > No, I mean the intention of this struct is that it _should_ fail

Yes my bad completely skipped over the implementation for the Cbor trait

File: vm/actor/src/builtin/puppet/mod.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+use crate::check_empty_params;
+use address::Address;
+use encoding::{tuple::*, Cbor, CodecProtocol, Error};
+use ipld_blockstore::BlockStore;
+use num_bigint::bigint_ser;
+use runtime::{ActorCode, Runtime};
+use serde::{Deserialize, Serialize};
+use vm::{ActorError, ExitCode, MethodNum, Serialized, TokenAmount, METHOD_CONSTRUCTOR};
+
```
Comment (ec2) [inline vm/actor/src/builtin/puppet/mod.rs:15]: Lol i know this actor does nothing, but for the sake of consistancy with the other actors, could you add the commit hash this was ported from?

File: vm/actor/src/builtin/puppet/mod.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    SendMethod = 2,
```
Comment (ec2) [inline vm/actor/src/builtin/puppet/mod.rs:20]: I believe this is just called Send in specs-actor

File: vm/actor/src/builtin/puppet/mod.rs [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            )
+        })?;
+
+        Ok(SendReturn {
+            return_bytes,
+            code: ExitCode::Ok,
+        })
+    }
+
+    fn send_failure<BS, RT>(rt: &mut RT, params: SendParams) -> Result<SendReturn, ActorError>
```
Comment (ec2) [inline vm/actor/src/builtin/puppet/mod.rs:108]: I don't remember if i specifically look for this and its definitely a nitpick, but it'd be nice if we followed the name of these methods in specs-actor exactly

File: vm/actor/src/builtin/puppet/mod.rs (hunk 4 above)
Comment (nannick) [reply vm/actor/src/builtin/puppet/mod.rs:20]: Ahh my bad I thought there wouldve been some issue cuz of the Send trait, just checked and guess its fine for enum varaints

File: vm/actor/src/builtin/puppet/mod.rs [hunk 6]
```diff
@@ -41,7 +41,21 @@ pub struct SendReturn {
...
+        })
+    }
+
+    fn unmarshal_cbor(_bz: &[u8]) -> Result<Self, Error> {
+        Err(Error::Unmarshalling {
+            description: "Automatic fail to Unmarshal".to_string(),
+            protocol: CodecProtocol::Cbor,
+        })
+    }
+}
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:60]: can you instead just have the serialize and deserialize impls fail? I'm pretty sure the cbor trait is used for all use cases, but there might be some serializations using the serde trait, and that would pass (just trying to future proof) the cbor implementation uses the serde impls by the way so this can be removed.

File: vm/actor/src/builtin/puppet/mod.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    ReturnMarshalCBORFailure = 4,
+    RuntimeTransactionMarshalCBORFailure = 5,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendParams {
+    pub to: Address,
+    #[serde(with = "bigint_ser")]
+    pub value: TokenAmount,
+    pub method: u64,
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:36]: MethodNum from the vm crate should probably be used to future proof

File: vm/actor/src/builtin/puppet/mod.rs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    pub params: Vec<u8>,
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:37]: This type should be `Serialized`. This will not marshall/unmarshall correctly (array of u8 rather than bytes)

File: vm/actor/src/builtin/puppet/mod.rs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        let res = rt.send(
+            params.to,
+            params.method,
+            Serialized::serialize(params.params).unwrap(),
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:104]: you should just use the params, this would reserialize the serialized bytes as serialized bytes lol

File: vm/actor/src/builtin/puppet/mod.rs [hunk 10]
```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:6]:
```suggestion
```

File: vm/actor/src/builtin/puppet/mod.rs [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+use crate::check_empty_params;
+use address::Address;
+use encoding::{tuple::*, Cbor};
+use ipld_blockstore::BlockStore;
+use num_bigint::bigint_ser;
+use runtime::{ActorCode, Runtime};
+use serde::de::{self, Deserializer};
+use serde::ser::{self, Serializer};
+use serde::{Deserialize, Serialize};
+
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:16]:
```suggestion
```

File: vm/actor/src/builtin/puppet/mod.rs [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            st.opt_fail = vec![];
+        })?;
+
+        Ok(())
+    }
+}
+
+impl ActorCode for Actor {
+    /// Invokes method with runtime on the actor's code. Method number will match one
+    /// defined by the Actor, and parameters will be serialized and used in execution
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:177]:
```suggestion
```
Redundant

File: vm/actor/src/builtin/puppet/mod.rs [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+            ActorError::new(
+                ExitCode::ErrIllegalState,
+                "failed to get return bytes".to_string(),
+            )
+        })?;
+
+        Ok(SendReturn {
+            return_bytes,
+            code: ExitCode::Ok,
+        })
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:118]:
```suggestion
        match res {
            Ok(return_bytes) => Ok(SendReturn {
                return_bytes: Some(return_bytes),
                code: ExitCode::Ok,
            }),
            Err(e) => Ok(SendReturn {
                return_bytes: None,
                code: e.exit_code(),
            }),
        }
```
You also need to change the return_bytes field to be an option, the go type can (and is nil in cases where there was an error)

File: vm/actor/src/builtin/puppet/mod.rs [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        let res = rt.send(
+            params.to,
+            params.method,
+            Serialized::serialize(FailToMarshalCBOR::default()).unwrap(),
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:134]:
```suggestion
            Serialized::serialize(FailToMarshalCBOR::default())?,
```

File: vm/actor/src/builtin/puppet/mod.rs [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            ActorError::new(
+                ExitCode::ErrIllegalState,
+                "failed to get return bytes".to_string(),
+            )
+        })?;
+
+        Ok(SendReturn {
+            return_bytes,
+            code: ExitCode::Ok,
+        })
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:148]:
```suggestion
        match res {
            Ok(return_bytes) => Ok(SendReturn {
                return_bytes: Some(return_bytes),
                code: ExitCode::Ok,
            }),
            Err(e) => Ok(SendReturn {
                return_bytes: None,
                code: e.exit_code(),
            }),
        }
```

File: vm/actor/src/builtin/puppet/mod.rs [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+
+    fn runtime_transaction_marshal_cbor_failure<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        rt.transaction(|st: &mut State, _| {
+            st.opt_fail = vec![];
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:168]: I don't think this fails in our case (and hope this doesn't ever get hit in our node outside of this, but I don't see why theirs would fail if that type never gets serialized) You might want to write a test or something to check if serializing an empty vec of `FailToMarshalCBOR` will error

File: vm/actor/src/builtin/puppet/mod.rs [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+                Ok(Serialized::serialize(res)?)
+            }
+
+            Some(Method::RuntimeTransactionMarshalCBORFailure) => {
+                check_empty_params(params)?;
+                Self::runtime_transaction_marshal_cbor_failure(rt)?;
+                Ok(Serialized::default())
+            }
+
+            _ => Err(rt.abort(ExitCode::SysErrInvalidMethod, "Invalid method")),
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:214]:
```suggestion
            None => Err(actor_error!(SysErrInvalidMethod; "Invalid method")),
```

File: vm/actor/src/builtin/puppet/mod.rs [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+            Some(Method::SendMarshalCBORFailure) => {
+                let res = Self::send_marshal_cbor_failure(rt, params.deserialize()?)?;
+                Ok(Serialized::serialize(res)?)
+            }
+            Some(Method::ReturnMarshalCBORFailure) => {
+                check_empty_params(params)?;
+                let res = Self::return_marshal_cbor_failure(rt)?;
+                Ok(Serialized::serialize(res)?)
+            }
+
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:207]:
```suggestion
```
nit

File: vm/actor/src/builtin/puppet/mod.rs (hunk 16 above)
Comment (nannick) [reply vm/actor/src/builtin/puppet/mod.rs:168]: Just wrote the test and a empty vec of FailToMarshalCBOR can be successfully serialized

File: vm/actor/src/builtin/puppet/mod.rs [hunk 19]
```diff
@@ -204,14 +197,12 @@ impl ActorCode for Actor {
...
             }
-
             Some(Method::RuntimeTransactionMarshalCBORFailure) => {
                 check_empty_params(params)?;
                 Self::runtime_transaction_marshal_cbor_failure(rt)?;
                 Ok(Serialized::default())
             }
-
-            _ => Err(rt.abort(ExitCode::SysErrInvalidMethod, "Invalid method")),
+            None => Err(rt.abort(ExitCode::SysErrInvalidMethod, "Invalid method")),
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/mod.rs:205]: can you also remove the rt.abort as I commented in favour of the macro? I want to remove this function after all actors are updated

File: vm/actor/tests/puppet_actor_test.rs [hunk 20]
```diff
@@ -72,5 +72,18 @@ fn simple_send() {
...
+fn serialize_test() {
+    let mut v: Vec<FailToMarshalCBOR> = vec![];
+
+    // Should pass becuase vec is empty
+    assert!(Serialized::serialize(&v).is_ok());
+
+    v.push(FailToMarshalCBOR::default());
+
+    // Should fail becuase vec is no longer empty
+    assert!(Serialized::serialize(v).is_err());
```
Comment (austinabell) [inline vm/actor/tests/puppet_actor_test.rs:88]: Can you also check the case for `runtime_transaction_marshal_cbor_failure`? my intuition is that it won't error as it's expected to

File: vm/actor/tests/puppet_actor_test.rs [hunk 21]
```diff
@@ -86,4 +86,17 @@ fn serialize_test() {
 
     // Should fail becuase vec is no longer empty
     assert!(Serialized::serialize(v).is_err());
+
+    let mut v: Vec<Option<FailToMarshalCBOR>> = vec![];
+
+    assert!(Serialized::serialize(&v).is_ok());
```
Comment (austinabell) [inline vm/actor/tests/puppet_actor_test.rs:92]: yeah this proves that for the transaction marshal failure call will not fail during serialization

File: vm/actor/src/builtin/puppet/state.rs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+impl Cbor for FailToMarshalCBOR {}
+
+#[derive(Default, Serialize_tuple, Deserialize_tuple)]
+pub struct State {
+    // OptFailToMarshalCBOR is to be used as an Option<T>, with T
+    // specialized to FailToMarshalCBOR. If the slice contains no values, the
+    // State struct will serialize/deserialize without issue. If the slice contains
+    // more than zero values, serialization/deserialization will fail.
+    pub opt_fail: Vec<Option<FailToMarshalCBOR>>,
```
Comment (austinabell) [inline vm/actor/src/builtin/puppet/state.rs:42]: This doesn't need to be wrapped in an `Option`. There is no case when you would need to denote the None variant
